B-Sockets-ServerTesting/sockets-servertesting.md

```diff
@@ -0,0 +1,296 @@
+requests:
+
+- The client sends an **HTTP request** for a resource on the server
+- The server sends an **HTTP response**
+
+**HTTP request**
+- First line: method, request URI, version
+    - Ex: "GET /index.html HTTP/1.0\r\n"
+- Followed by 0 or more headers
+    - Ex: "Host: www.google.com\r\n"
+- Followed by an empty line
+    - "\r\n"
+
+**HTTP response**
+- First line: response status
+    - Success: "HTTP/1.0 200 OK\r\n"
+    - Failure: "HTTP/1.0 404 Not Found\r\n"
+- Followed by 0 or more response headers
+- Followed by an empty line
+- Followed by the content of the response
+    - Ex: image file or HTML file
+
+We can see the contents of real HTTP requests using `netcat` by pretending to be
+either a web client or server. 
+
+Let's first act as a web server. We tell `netcat` to open a server connection
+with `nc -l 10000`, and then in a web browser navigate to the URL with the
+domain name of this server. We can use the domain name `localhost` to specify
+the local computer rather than connecting to a remote computer over the
+internet. In Chrome, we'll navigate to the URL
+`http://localhost:10000/index.html`. `netcat` outputs this:
+
+```
+$ nc -l 10000
+GET /index.html HTTP/1.1   # GET == method; /index.html == request URI; HTTP/1.1 == version
+Host: localhost:10000      # header
+Connection: keep-alive     # more headers...
+-removed for brevity-
+                           # blank newline to indicate end of headers/request
+```
+
+To act as a client, we can type our HTTP request manually into netcat rather
+than doing it through the web browser. Here, we try to send an HTTP request to
+the domain name `example.com` on port `80` (the default for HTTP web servers)
+for the URI `/index.html`. Note that we specify the `-C` with `netcat` so that
+newlines are `\r\n` rather than `\n` -- a requirement of the HTTP protocol. This
+flag may vary depending on `netcat` version.
+
+```
+$ nc -C example.com 80
+GET /index.html HTTP/1.0    # GET == method; /index.html == request URI; HTTP/1.1 == version
+                            # blank line to specify end of request
+HTTP/1.0 200 OK             # start of HTTP response. HTTP/1.0 == version; 200 OK == response status
+Accept-Ranges: bytes        # header
+Content-Type: text/html     # more headers...
+-removed for brevity-
+                            # blank newline to indicate end of headers and start of file contents
+<!doctype html>             # HTML contents
+<html>
+<head>
+    <title>Example Domain</title>
+-removed for brevity-
+```
+
+Testing your multi-server
+---------------------------
+### Siege
+Siege is a command-line tool that allows you to benchmark your webserver using
+load testing. Given a few parameters, Siege gives you information about the
+number of successful transactions to your website, percent availability, the
+latency/throughput of your server, and more.
+
+To install siege, run the following command:
+
+```sudo apt install siege```
```

Comment:
this should be

\`\`\`
sudo ...
\`\`\`

same with the next two siege commands below